File: packages-gen/js/substrate/page-util.js
const xtd = require("@xtalk/lang/common-data.js")

const event_model = require("@xtalk/event/base-model.js")

function wrap_space_args(handler){
  return function (context){
    let args = context["args"] || [];
    let params = [context];
    xtd.arr_assign(params,args);
    return handler.apply(null,params);
  };
}

function check_event(pred,signal,event,ctx){
  let check = false;
  try{
    let t = null;
    if(null == pred){
      t = true;
    }
    else if("boolean" == (typeof pred)){
      t = pred;
    }
    else if("function" == (typeof pred)){
      t = pred(signal,ctx);
    }
    else if((null != pred) && ("object" == (typeof pred)) && !Array.isArray(pred)){
      t = pred[signal];
    }
    else{
      t = (signal == pred);
    }
    if(true == t){
      check = true;
    }
    else if("function" == (typeof t)){
      check = t.apply(null,[event,ctx]);
    }
  }
  catch(err){
    check = false;
  }
  return check;
}

function run_tail_call(context,refresh_deps_fn){
  let acc = context["acc"];
  let path = context["path"];
  let node = context["node"];
  let space_id = (context["space"])["id"];
  let group_id = path[0];
  let model_id = path[1];
  if(acc && !acc["error"] && refresh_deps_fn){
    return Promise.resolve().then(function (){
      return refresh_deps_fn(node,space_id,group_id,model_id,refresh_deps_fn);
    }).then(function (_){
      return acc;
    });
  }
  else{
    return acc;
  }
}

function run_remote(context,save_output,path,refresh_deps_fn){
  context["acc"]["path"] = path;
  return event_model.pipeline_run_remote(context,save_output,event_model.async_fn_promise,null,null).then(function (_){
    return run_tail_call(context,refresh_deps_fn);
  });
}

function run_refresh(context,disabled,path,refresh_deps_fn){
  context["acc"]["path"] = path;
  return event_model.pipeline_run(context,disabled,event_model.async_fn_promise,null,null).then(function (_){
    return run_tail_call(context,refresh_deps_fn);
  });
}

function get_group_deps(group_id,models){
  let all_deps = {};
  for(let [model_id,model_entry] of Object.entries(models)){
    let deps = model_entry["deps"];
    for(let path of deps || []){
      path = (Array.isArray(path) ? path : [group_id,path]);
      xtd.set_in(all_deps,[path[0],path[1],model_id],true);
    };
  };
  return all_deps;
}

function raw_callback_id(space_id){
  return "@/raw/page/" + (space_id || "");
}

function register_page_trigger(node,signal,trigger_fn,meta){
  let entry = {"id":signal,"fn":trigger_fn,"meta":meta || {}};
  node["triggers"][signal] = entry;
  return entry;
}

function unregister_page_trigger(node,signal){
  let triggers = node["triggers"];
  let prev = triggers[signal];
  delete(triggers[signal]);
  return prev;
}

module.exports = {
  ["wrap_space_args"]:wrap_space_args,
  ["check_event"]:check_event,
  ["run_tail_call"]:run_tail_call,
  ["run_remote"]:run_remote,
  ["run_refresh"]:run_refresh,
  ["get_group_deps"]:get_group_deps,
  ["raw_callback_id"]:raw_callback_id,
  ["register_page_trigger"]:register_page_trigger,
  ["unregister_page_trigger"]:unregister_page_trigger
}

File: packages-gen/dart/libs/xt-substrate/lib/page-util.dart
import 'package:xtalk_lang/common-data.dart' as xtd;
import 'package:xtalk_event/base-model.dart' as event_model;
import 'dart:async';



wrap_space_args(handler) {
  return (context) {
    var args = context["args"] ?? <dynamic>[];
    var params = <dynamic>[context];
    xtd.arr_assign(params,args);
    return Function.apply(handler,params);
  };
}

check_event(pred, signal, event, ctx) {
  var check = false;
  try{
    var t = null;
    if(null == pred){
      t = true;
    }
    else if("bool" == (pred.runtimeType).toString()){
      t = pred;
    }
    else if((pred.runtimeType).toString().contains("Function") || (pred.runtimeType).toString().contains("=>") || (pred).toString().startsWith("Closure")){
      t = pred(signal,ctx);
    }
    else if(("Map" == (pred.runtimeType).toString()) || (pred.runtimeType).toString().startsWith("_Map") || (pred.runtimeType).toString().startsWith("LinkedMap")){
      t = pred[signal];
    }
    else{
      t = (signal == pred);
    }
    if(true == t){
      check = true;
    }
    else if((t.runtimeType).toString().contains("Function") || (t.runtimeType).toString().contains("=>") || (t).toString().startsWith("Closure")){
      check = Function.apply(t,<dynamic>[event,ctx]);
    }
  }
  catch(err){
    check = false;
  }
  return check;
}

run_tail_call(context, refresh_deps_fn) {
  var acc = context["acc"];
  var path = context["path"];
  var node = context["node"];
  var space_id = (context["space"])["id"];
  var group_id = path[0];
  var model_id = path[1];
  if(((null != acc) && (false != acc)) && !(() {
    var dart_truthy__51430 = acc["error"];
    return (null != dart_truthy__51430) && (false != dart_truthy__51430);
  })() && ((null != refresh_deps_fn) && (false != refresh_deps_fn))){
    return ((Future.sync(() => Future.sync(() {
      return Function.apply(
        (refresh_deps_fn as Function),
        <dynamic>[node,space_id,group_id,model_id,refresh_deps_fn]
      );
    }))) as Future<dynamic>).then((value) async { return await Function.apply((_) {
      return acc;
    },<dynamic>[value]); });
  }
  else{
    return acc;
  }
}

run_remote(context, save_output, path, refresh_deps_fn) {
  context["acc"]["path"] = path;
  return ((Future.sync(() => event_model.pipeline_run_remote(context,save_output,event_model.async_fn_promise,null,null))) as Future<dynamic>).then((value) async { return await Function.apply((_) {
    return run_tail_call(context,refresh_deps_fn);
  },<dynamic>[value]); });
}

run_refresh(context, disabled, path, refresh_deps_fn) {
  context["acc"]["path"] = path;
  return ((Future.sync(() => event_model.pipeline_run(context,disabled,event_model.async_fn_promise,null,null,null))) as Future<dynamic>).then((value) async { return await Function.apply((_) {
    return run_tail_call(context,refresh_deps_fn);
  },<dynamic>[value]); });
}

get_group_deps(group_id, models) {
  var all_deps = <dynamic, dynamic>{};
  for(var entry_51431 in models.entries){
    var model_id = entry_51431.key;
    var model_entry = entry_51431.value;
    var deps = model_entry["deps"];
    var arr_51432 = deps ?? <dynamic>[];
    for(var i51433 = 0; i51433 < arr_51432.length; ++i51433){
      var path = arr_51432[i51433];
      path = (((path.runtimeType).toString().startsWith("List") || (path.runtimeType).toString().startsWith("_GrowableList")) ? path : <dynamic>[group_id,path]);
      xtd.set_in(all_deps,<dynamic>[path[0],path[1],model_id],true);
    };
  };
  return all_deps;
}

raw_callback_id(space_id) {
  return "@/raw/page/" + (space_id ?? "");
}

register_page_trigger(node, signal, trigger_fn, meta) {
  var entry = <dynamic, dynamic>{
    "id":signal,
    "fn":trigger_fn,
    "meta":meta ?? <dynamic, dynamic>{}
  };
  node["triggers"][signal] = entry;
  return entry;
}

unregister_page_trigger(node, signal) {
  var triggers = node["triggers"];
  var prev = triggers[signal];
  triggers.remove(signal);
  return prev;
}